plugins/auth-backend-module-pinniped-provider/src/authenticator.ts

```diff
@@ -0,0 +1,167 @@
+    const issuer = await Issuer.discover(
+      `${config.getString(
+        'federationDomain',
+      )}/.well-known/openid-configuration`,
+    );
+    const client = new issuer.Client({
+      access_type: 'offline', // this option must be passed to provider to receive a refresh token
+      client_id: config.getString('clientId'),
+      client_secret: config.getString('clientSecret'),
+      redirect_uris: [callbackUrl],
+      response_types: ['code'],
+      scope: config.getOptionalString('scope') || '',
+      id_token_signed_response_alg: 'ES256',
+    });
+    const strategy = new OidcStrategy(
+      {
+        client,
+        passReqToCallback: false,
+      },
+      (
+        tokenset: TokenSet,
+        done: PassportDoneCallback<
+          { tokenset: TokenSet },
+          {
+            refreshToken?: string;
+          }
+        >,
+      ) => {
+        done(undefined, { tokenset }, {});
+      },
+    );
+
+    return { strategy, client };
+  },
+
+  async start(input, implementation) {
+    const { strategy } = await implementation;
+    const stringifiedAudience = input.req.query?.audience as string;
+    const decodedState = decodeOAuthState(input.state);
+    const state = { ...decodedState, audience: stringifiedAudience };
+    const options: Record<string, string> = {
+      scope:
+        input.scope ||
+        'openid pinniped:request-audience username offline_access',
+      state: encodeOAuthState(state),
+    };
+
+    return new Promise((resolve, reject) => {
+      strategy.redirect = (url: string) => {
+        resolve({ url });
+      };
+      strategy.error = (error: Error) => {
+        reject(error);
+      };
+      strategy.authenticate(input.req, { ...options });
+    });
+  },
+
+  async authenticate(input, implementation) {
+    const { strategy, client } = await implementation;
+    const { req } = input;
+    const { searchParams } = new URL(req.url, 'https://pinniped.com');
+    const stateParam = searchParams.get('state');
+    const audience = stateParam
+      ? decodeOAuthState(stateParam).audience
+      : undefined;
+
+    return new Promise((resolve, reject) => {
+      strategy.success = user => {
+        (audience
+          ? client
+              .grant({
+                grant_type: 'urn:ietf:params:oauth:grant-type:token-exchange',
+                subject_token: user.tokenset.access_token,
+                audience,
+                subject_token_type:
+                  'urn:ietf:params:oauth:token-type:access_token',
+                requested_token_type: 'urn:ietf:params:oauth:token-type:jwt',
+              })
+              .then(tokenset => tokenset.access_token)
+              .catch(err =>
+                reject(
+                  new Error(
+                    `Failed to get cluster specific ID token for "${audience}", RFC8693 token exchange failed with error: ${err}`,
+                  ),
+                ),
+              )
```

Comment:
Could we extract this bit to a private helper method with a nice keyword-filled name, like:
```suggestion
          ? this.rfc8693TokenExchange({ subject_token: user.tokenset.access_token, audience })
```
It might be easier to understand if we have to come back to this in the future

Reply:
Sure thing. Done.